tests/pjit_test.py

```diff
@@ -0,0 +1,173 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from contextlib import contextmanager
+from functools import partial
+from typing import Generator, List, Tuple
+from unittest import SkipTest
+
+from absl.testing import absltest
+import numpy as np
+
+import jax
+import jax.numpy as jnp
+from jax import test_util as jtu
+# TODO(skye): do we still wanna call this PartitionSpec?
+from jax.experimental import PartitionSpec as P
+from jax.experimental.maps import mesh
+from jax.experimental.pjit import pjit, with_sharding_constraint
+from jax.interpreters import pxla
+from jax._src.util import unzip2, prod
+
+from jax.config import config
+config.parse_flags_with_absl()
+
+# TODO(skye): move into test_util and dedup with xmap_test.py
+MeshSpec = List[Tuple[str, int]]
+
+@contextmanager
+def with_mesh(named_shape: MeshSpec) -> Generator[None, None, None]:
+  """Test utility for setting up meshes given mesh data from `schedules`."""
+  # This is similar to the `with_mesh` function above, but isn't a decorator.
+  axis_names, shape = unzip2(named_shape)
+  size = prod(shape)
+  local_devices = list(jax.local_devices())
+  if len(local_devices) < size:
+    raise SkipTest(f"Test requires {size} local devices")
+  mesh_devices = np.array(local_devices[:size]).reshape(shape)
+  with mesh(mesh_devices, axis_names):
+    yield
+
+# TODO(skye): make the buffer donation utils part of JaxTestCase
+class PJitTest(jtu.BufferDonationTestCase):
+
+  @with_mesh([('x', 2)])
+  def testBasic1D(self):
+    @partial(pjit,
+             in_axis_resources=(P('x'), P('x')),
+             out_axis_resources=None)
+    def f(x, y):
+      return x + y
+
+    shape = (8, 8)
+    x = np.arange(prod(shape), dtype=np.float32).reshape(shape)
+    actual = f(x, x + 1)
+    expected = x + (x + 1)
+    self.assertAllClose(actual, expected, check_dtypes=False)
+    self.assertIsInstance(actual, pxla.ShardedDeviceArray)
+    self.assertLen(actual.device_buffers, 2)
+    self.assertAllClose(actual.device_buffers[0].to_py(), expected,
+                        check_dtypes=False)
+
+  @with_mesh([('x', 2), ('y', 2)])
+  def testBasic2D(self):
+    @partial(pjit,
+             in_axis_resources=(P(None, 'x', 'y'), P('y')),
+             out_axis_resources=P('x'))
+    def f(x, y):
+      return x @ y
+
+    x_shape = (8, 6, 4)
+    y_shape = (4, 2)
+    x = jnp.arange(np.prod(x_shape)).reshape(x_shape)
+    y = jnp.arange(np.prod(y_shape)).reshape(y_shape)
+    actual = f(x, y)
+    expected = x @ y
+    self.assertAllClose(actual, expected, check_dtypes=False)
+    self.assertIsInstance(actual, pxla.ShardedDeviceArray)
+    self.assertLen(actual.device_buffers, 4)
+
+    split0, split1 = np.split(expected, 2)
+    self.assertAllClose(actual.device_buffers[0].to_py(), split0,
```

Comment:
I am wondering if it is specified that the device_buffers will be assigned as tested here?

Reply:
I don't think this is part of the public API, no. A less implementation-specific way to do this might be to look at each device_buffer's device to determine which split it should have, although that might make the test hard to reason about. I will probably focus on other things first before revisiting this test, but something to keep in mind if/when we do revisit it (e.g. if it breaks due to this detail :wink: ).